### PR TITLE
introduce CaseBlock as a block-scoped container for switch statements

### LINKED ISSUE #2280: Use-before-def is reported incorrectly for switch statement expressions
```TypeScript
var x = 10;

switch (x) {
    case 10:
        let x = 20;
}
```

It is worth ensuring the spec behavior, but I was under the impression that the `x` in `switch (x)` would refer to the outer `x`. Instead, we get

> Block-scoped variable 'x' used before its declaration.

### PR DESCRIPTION
**Pull request by vladima**
Fixes #2280

### CODE REVIEW

File: src/services/formatting/smartIndenter.ts [hunk 1]
```diff
@@ -357,7 +357,7 @@ module ts.formatting {
                 case SyntaxKind.ModuleBlock:
                 case SyntaxKind.ObjectLiteralExpression:
                 case SyntaxKind.TypeLiteral:
-                case SyntaxKind.SwitchStatement:
+                case SyntaxKind.CaseBlock:
```
Comment (DanielRosenwasser) [inline src/services/formatting/smartIndenter.ts]: `CaseClause`

File: src/compiler/types.ts [hunk 2]
```diff
@@ -790,6 +791,10 @@ module ts {
 
     export interface SwitchStatement extends Statement {
         expression: Expression;
+        caseBlock: CaseBlock;
+    }
+
+    export interface CaseBlock extends Node {
```
Comment (DanielRosenwasser) [inline src/compiler/types.ts]: Can we just add a comment that this is more like a SwitchBlock? It's somewhat misleading

**Comment by DanielRosenwasser** [comment]
:+1:

File: src/compiler/emitter.ts [hunk 3]
```diff
@@ -3541,10 +3541,10 @@ module ts {
                 write(" ");
                 emitToken(SyntaxKind.OpenBraceToken, endPos);
```
Comment (CyrusNajmabadi) [inline src/compiler/emitter.ts]: Nit: can we have an emitCaseBlock function where you do all the appropriate work for that child node specifically?

**Pull request merged**